File: shared/src/commonMain/kotlin/opensignal/risk/RiskEngine.kt
package opensignal.risk

import kotlin.math.abs
import kotlin.math.max
import opensignal.models.LevelKind
import opensignal.models.RiskManagementPlan
import opensignal.models.RiskParameters
import opensignal.models.TechnicalAnalysis
import opensignal.models.TradeOption
import opensignal.models.TradePlan
import opensignal.models.TradeSide

data class RiskEngineResult(
    val plan: TradePlan,
    val risk: RiskManagementPlan
)

class RiskEngine {

    fun buildPlan(technical: TechnicalAnalysis, parameters: RiskParameters): RiskEngineResult {
        val supports = technical.supportResistance
            .filter { it.kind == LevelKind.SUPPORT }
            .sortedByDescending { it.strength }
        val resistances = technical.supportResistance
            .filter { it.kind == LevelKind.RESISTANCE }
            .sortedByDescending { it.strength }

        val current = technical.currentPrice
        val baseRisk = parameters.accountBalance * (parameters.riskPerTradePercent / 100.0)

        val strongestSupport = supports.firstOrNull()?.price ?: current * 0.995
        val strongestResistance = resistances.firstOrNull()?.price ?: current * 1.005

        val buyPrimary = buildOption(
            side = TradeSide.BUY,
            entry = current,
            stop = minOf(strongestSupport, current * 0.995),
            target = max(
                strongestResistance,
                current + abs(current - strongestSupport) * parameters.minimumRiskReward
            ),
            confidence = technical.trendConfidence,
            baseRisk = baseRisk,
            leverage = parameters.leverage,
            note = "Trend-following buy from current price"
        )

        val buySecondary = buildOption(
            side = TradeSide.BUY,
            entry = strongestSupport,
            stop = strongestSupport * 0.997,
            target = max(
                strongestResistance,
                strongestSupport + abs(strongestSupport - strongestSupport * 0.997) *
                    (parameters.minimumRiskReward + 0.5)
            ),
            confidence = (technical.trendConfidence * 0.92).coerceAtLeast(0.35),
            baseRisk = baseRisk,
            leverage = parameters.leverage,
            note = "Pullback buy from strongest support"
        )

        val sellPrimary = buildOption(
            side = TradeSide.SELL,
            entry = current,
            stop = max(strongestResistance, current * 1.005),
            target = minOf(
                strongestSupport,
                current - abs(strongestResistance - current) * parameters.minimumRiskReward
            ),
            confidence = technical.trendConfidence,
            baseRisk = baseRisk,
            leverage = parameters.leverage,
            note = "Trend-following sell from current price"
        )

        val sellSecondary = buildOption(
            side = TradeSide.SELL,
            entry = strongestResistance,
            stop = strongestResistance * 1.003,
            target = minOf(
                strongestSupport,
                strongestResistance - abs(strongestResistance * 1.003 - strongestResistance) *
                    (parameters.minimumRiskReward + 0.5)
            ),
            confidence = (technical.trendConfidence * 0.92).coerceAtLeast(0.35),
            baseRisk = baseRisk,
            leverage = parameters.leverage,
            note = "Rejection sell from strongest resistance"
        )

        val options = listOf(buyPrimary, buySecondary, sellPrimary, sellSecondary)
        val warnings = mutableListOf<String>()
        options.filter { it.riskReward < parameters.minimumRiskReward }.forEach {
            warnings += "${it.side} option at ${it.entry} is below minimum R:R"
        }

        if (parameters.riskPerTradePercent > 2.0) {
            warnings += "Risk per trade above 2% is aggressive for volatile assets"
        }

        val riskPlan = RiskManagementPlan(
            parameters = parameters,
            warnings = warnings,
            maxLossPerTrade = baseRisk,
            maxTotalPortfolioRisk = baseRisk * parameters.maxOpenTrades
        )

        return RiskEngineResult(
            plan = TradePlan(
                buyOptions = listOf(buyPrimary, buySecondary),
                sellOptions = listOf(sellPrimary, sellSecondary)
            ),
            risk = riskPlan
        )
    }

    private fun buildOption(
        side: TradeSide,
        entry: Double,
        stop: Double,
        target: Double,
        confidence: Double,
        baseRisk: Double,
        leverage: Double,
        note: String
    ): TradeOption {
        val stopDistance = abs(entry - stop).coerceAtLeast(entry * 0.0005)
        val rewardDistance = abs(target - entry).coerceAtLeast(entry * 0.0005)
        val rr = rewardDistance / stopDistance
        val positionUnits = (baseRisk / stopDistance) * leverage

        return TradeOption(
            side = side,
            entry = entry,
            stopLoss = stop,
            takeProfit = target,
            riskReward = rr,
            positionSizeUnits = positionUnits,
            capitalAtRisk = baseRisk,
            confidence = confidence,
            note = note
        )
    }
}

File: ai-service/signal_engine/risk_engine.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List


@dataclass
class RiskInput:
    current_price: float
    levels: list[dict]
    trend_confidence: float
    account_balance: float
    risk_per_trade_percent: float
    leverage: float
    minimum_rr: float


def _build_option(
    side: str,
    entry: float,
    stop_loss: float,
    take_profit: float,
    confidence: float,
    risk_capital: float,
) -> dict:
    stop_distance = max(abs(entry - stop_loss), entry * 0.0005)
    reward_distance = max(abs(take_profit - entry), entry * 0.0005)
    rr = reward_distance / stop_distance
    position_units = risk_capital / stop_distance

    return {
        "side": side,
        "entry": round(entry, 4),
        "stop_loss": round(stop_loss, 4),
        "take_profit": round(take_profit, 4),
        "risk_reward": round(rr, 4),
        "position_size_units": round(position_units, 4),
        "capital_at_risk": round(risk_capital, 4),
        "confidence": round(confidence, 3),
        "note": "Generated from screenshot structure and liquidity model outputs",
    }


def generate_trade_plan(inp: RiskInput) -> dict:
    support_levels = [l for l in inp.levels if l["kind"] == "SUPPORT"]
    resistance_levels = [l for l in inp.levels if l["kind"] == "RESISTANCE"]

    support = support_levels[0]["price"] if support_levels else inp.current_price * 0.992
    resistance = resistance_levels[0]["price"] if resistance_levels else inp.current_price * 1.008

    risk_capital = inp.account_balance * (inp.risk_per_trade_percent / 100.0)

    buy_primary = _build_option(
        side="BUY",
        entry=inp.current_price,
        stop_loss=min(support, inp.current_price * 0.995),
        take_profit=max(resistance, inp.current_price + abs(inp.current_price - support) * inp.minimum_rr),
        confidence=inp.trend_confidence,
        risk_capital=risk_capital,
    )
    buy_pullback = _build_option(
        side="BUY",
        entry=support,
        stop_loss=support * 0.997,
        take_profit=max(resistance, support + abs(support * 0.003) * (inp.minimum_rr + 0.5)),
        confidence=max(0.35, inp.trend_confidence * 0.92),
        risk_capital=risk_capital,
    )

    sell_primary = _build_option(
        side="SELL",
        entry=inp.current_price,
        stop_loss=max(resistance, inp.current_price * 1.005),
        take_profit=min(support, inp.current_price - abs(resistance - inp.current_price) * inp.minimum_rr),
        confidence=inp.trend_confidence,
        risk_capital=risk_capital,
    )
    sell_rejection = _build_option(
        side="SELL",
        entry=resistance,
        stop_loss=resistance * 1.003,
        take_profit=min(support, resistance - abs(resistance * 0.003) * (inp.minimum_rr + 0.5)),
        confidence=max(0.35, inp.trend_confidence * 0.92),
        risk_capital=risk_capital,
    )

    options = [buy_primary, buy_pullback, sell_primary, sell_rejection]
    warnings: List[str] = []
    for option in options:
        if option["risk_reward"] < inp.minimum_rr:
            warnings.append(f"{option['side']} option at {option['entry']} below min RR")

    if inp.risk_per_trade_percent > 2.0:
        warnings.append("Risk per trade above 2% is aggressive")

    if inp.leverage > 5.0:
        warnings.append("Leverage above 5x increases liquidation risk")

    return {
        "buy_options": [buy_primary, buy_pullback],
        "sell_options": [sell_primary, sell_rejection],
        "risk": {
            "max_loss_per_trade": round(risk_capital, 4),
            "max_total_portfolio_risk": round(risk_capital * 3, 4),
            "warnings": warnings,
        },
    }
